synthetic/image/poisson_blending.py

```diff
@@ -0,0 +1,224 @@
+    testimage = Image.new(mode='RGB', size=image_size, color=bg_color)
+    draw = ImageDraw.Draw(testimage)
+
+    c = 1.5
+    valid_sizes = []
+    while not valid_sizes and c < 100:
+        image_size_min = image_size[0] / c, image_size[1] / c
+
+        for font_size in fonts_by_size.keys():
+            tx, ty = draw.textsize(text=text, font=fonts_by_size[font_size])
+            if (tx < image_size[0]) and (image_size_min[1] < ty < image_size[1]):
+                valid_sizes.append(font_size)
+
+        c *= 1.5
+
+    if not valid_sizes:
+        raise Exception('Could not draw text on image')
+
+    fontsize = (min(valid_sizes) + max(valid_sizes)) // 2
+    font = fonts_by_size[fontsize]
+    tx, ty = draw.textsize(text=text, font=font)
+    xm, ym = [int(s / 2) for s in testimage.size]
+    x0, y0 = xm - tx / 2, ym - ty / 2
+    draw.text((x0, y0), text, font=font, fill=fg_color)
+    return testimage.rotate(angle, fillcolor=bg_color)
+
+
+def add_text_to_image(text, image_array, bounding_box, angle, fonts_by_size):
+    image_array = np.array(image_array)
+    xmin, ymin, xmax, ymax = bounding_box
+    dx, dy = xmax - xmin, ymax - ymin
+    bdy_ratio = 0.005
+    bdy_ratio_x, bdy_ratio_y = bdy_ratio * sqrt(dy / dx), bdy_ratio * sqrt(dx / dy)
+    bx, by = max(int(dx * bdy_ratio_x), 1), max(int(dy * bdy_ratio_y), 1)
+    xmb, ymb = [r - 2 * b for r, b in [(xmin, bx), (ymin, by)]]
+    xMb, yMb = [r + 2 * b for r, b in [(xmax, bx), (ymax, by)]]
+    xmb, xMb = [clip(v, 1, image_array.shape[1] - 1) for v in [xmb, xMb]]
+    ymb, yMb = [clip(v, 1, image_array.shape[0] - 1) for v in [ymb, yMb]]
+
+    image_mask = np.zeros((yMb - ymb, xMb - xmb), dtype=np.uint8)
+    image_mask[by:-by, bx:-bx] = 1
+
+    image_target = np.array(image_array[ymb:yMb, xmb:xMb, :])
+    bg_color, fg_color = get_bg_fg_cols(image_target)
+    image_source = text_to_image(text, image_target.shape[:2][::-1], fonts_by_size, bg_color, fg_color, angle)
+    image_ret = blend(image_target, np.asarray(image_source), image_mask)
+
+    image_array[ymb:yMb, xmb:xMb, :] = image_ret
+    return image_array
+
+
+def adjust_box(edginess, bounding_box):
+    r = 0.1
+    xmin, ymin, xmax, ymax = bounding_box
+    dx, dy = int(xmax - xmin), int(ymax - ymin)
+    dx, dy = int(r * sqrt(dy * dx)), int(r * sqrt(dy * dx))
+    dx, dy = max(dx, 5), max(dy, 5)
+    dxh, dyh = int(dx / 3), int(dy / 3)
+
+    min_val = 1E20
+    best_bounding_box = None
+    D = max(2, int(np.sqrt(dx * dy) / 10))
+    for possible_box in itertools.product(
+        range(xmin - dx, xmin + dxh + 1),
+        range(ymin - dy, ymin + dyh + 1),
+        range(xmax - dxh, xmax + dx + 1),
+        range(ymax - dxh, ymax + dy + 1),
+    ):
+        xm, ym, xM, yM = possible_box
+        e1 = edginess[ym - D:yM + D + 1, xm - D:xM + D + 1]
+        e2 = edginess[ym + D:yM - D + 1, xm + D:xM - D + 1]
+        E = np.sum(e1) - np.sum(e2)
+        if E < min_val:
```

Comment:
This is kind of like checking the gradient of the box, the lesser the gradient the better?